File: python/47_PermutationsII.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        total_set = []
        def helper(current_nums, current_set, count=0):
            if not current_nums:
                total_set.append(current_set)
            for i in range(len(current_nums)):
		if i == 0 or current_nums[i] != current_nums[i-1]:
                    helper(current_nums[0:i]+current_nums[i+1:],current_set+[current_nums[i]])
            return total_set
        return helper(nums, [], 0)

File: swift/47_PermutationsII.swift
class Solution {
    func permuteUnique(_ allNumbers: [Int]) -> [[Int]] {
        func getPermutations(_ candidate: [Int]) -> [[Int]] {
            var candidatePermutation = candidate
            if candidatePermutation.count == nums.count {
                allPermutations.append(candidatePermutation)
                return allPermutations
            }
            
            for i in 0..<nums.count {
                if nums[i] != -11 && (i == 0 || nums[i] != nums[i-1]) {
                    var temp = nums[i]
                    nums[i] = -11
                    candidatePermutation.append(temp)
                    getPermutations(candidatePermutation)
                    candidatePermutation.removeLast()
                    nums[i] = temp
                }
            }
            
            return allPermutations
        }
        var nums = allNumbers.sorted()
        var allPermutations = [[Int]]()
        return getPermutations([Int]())
    }
}
